File: python/src/hooks/distill_hook.py
#!/usr/bin/env python3
"""Distill hook handler for PreCompact and SessionEnd events.

Receives hook event data via stdin (JSON), extracts knowledge
from the session transcript, and stores it.

Usage: echo '{"session_id":"...","transcript_path":"..."}' | python -m src.hooks.distill_hook
"""

import asyncio
import json
import sys

from ..extractor.extractor import extract_knowledge
from ..store.metadata import MetadataStore
from ..store.scope import detect_project_root
from ..store.types import ExtractionTrigger
from ..store.vector import VectorStore


async def main() -> None:
    # Read stdin
    input_data = sys.stdin.read().strip()
    if not input_data:
        print("distill-hook: no input received on stdin", file=sys.stderr)
        sys.exit(1)

    try:
        hook_data = json.loads(input_data)
    except json.JSONDecodeError:
        print("distill-hook: invalid JSON on stdin", file=sys.stderr)
        sys.exit(1)

    session_id = hook_data.get("session_id")
    transcript_path = hook_data.get("transcript_path")

    if not session_id or not transcript_path:
        print(
            "distill-hook: missing session_id or transcript_path",
            file=sys.stderr,
        )
        sys.exit(1)

    # Determine trigger type from hook event name
    hook_event_name = hook_data.get("hook_event_name", "")
    trigger: ExtractionTrigger = (
        "pre_compact" if hook_event_name == "PreCompact" else "session_end"
    )

    cwd = hook_data.get("cwd")
    project_root = detect_project_root(cwd) if cwd else detect_project_root()
    project_name = project_root.split("/")[-1] if project_root else None

    print(
        f"distill-hook: extracting from session {session_id} ({trigger})",
        file=sys.stderr,
    )

    try:
        chunks = await extract_knowledge(
            transcript_path=transcript_path,
            session_id=session_id,
            trigger=trigger,
            project_name=project_name,
        )

        if not chunks:
            print("distill-hook: no knowledge extracted", file=sys.stderr)
            return

        saved = 0
        for chunk in chunks:
            try:
                meta = MetadataStore(chunk.scope, project_root or None)
                vector = VectorStore(chunk.scope, project_root or None)

                inserted = meta.insert(chunk)
                vector.index(inserted.id, inserted.content, inserted.tags)

                meta.close()
                vector.close()
                saved += 1
            except Exception as err:
                print(
                    f"distill-hook: failed to save chunk: {err}",
                    file=sys.stderr,
                )

        print(
            f"distill-hook: extracted {len(chunks)}, saved {saved} knowledge chunks",
            file=sys.stderr,
        )
    except Exception as err:
        print(f"distill-hook: extraction failed: {err}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: dart/lib/hooks/distill_hook.dart
import 'dart:convert';
import 'dart:io';
import '../extractor/extractor.dart';
import '../store/metadata.dart';
import '../store/vector.dart';
import '../store/scope.dart';
import '../store/types.dart';

/// Distill hook handler for PreCompact and SessionEnd events.
///
/// Receives hook event data via stdin (JSON), extracts knowledge
/// from the session transcript, and stores it.
///
/// Usage: echo '{"session_id":"...","transcript_path":"..."}' | dart run distill_hook.dart
Future<void> runHook() async {
  // Read stdin
  final input = await _readStdin();
  if (input.isEmpty) {
    stderr.writeln('distill-hook: no input received on stdin');
    exit(1);
  }

  Map<String, dynamic> hookData;
  try {
    hookData = jsonDecode(input) as Map<String, dynamic>;
  } catch (_) {
    stderr.writeln('distill-hook: invalid JSON on stdin');
    exit(1);
  }

  final sessionId = hookData['session_id'] as String?;
  final transcriptPath = hookData['transcript_path'] as String?;

  if (sessionId == null || transcriptPath == null) {
    stderr.writeln('distill-hook: missing session_id or transcript_path');
    exit(1);
  }

  // Determine trigger type from hook event name
  final hookEventName = hookData['hook_event_name'] as String?;
  final trigger = hookEventName == 'PreCompact'
      ? ExtractionTrigger.preCompact
      : ExtractionTrigger.sessionEnd;

  final cwd = hookData['cwd'] as String?;
  final projectRoot =
      cwd != null ? detectProjectRoot(cwd: cwd) : detectProjectRoot();
  final projectName = projectRoot?.split('/').last;

  stderr.writeln(
    'distill-hook: extracting from session $sessionId (${trigger.value})',
  );

  try {
    final chunks = await extractKnowledge(
      transcriptPath: transcriptPath,
      sessionId: sessionId,
      trigger: trigger,
      projectName: projectName,
    );

    if (chunks.isEmpty) {
      stderr.writeln('distill-hook: no knowledge extracted');
      return;
    }

    var saved = 0;
    for (final chunk in chunks) {
      try {
        final meta = MetadataStore(chunk.scope, projectRoot: projectRoot);
        final vector = VectorStore(chunk.scope, projectRoot: projectRoot);

        final inserted = meta.insert(chunk);
        vector.index(inserted.id, inserted.content, inserted.tags);

        meta.close();
        vector.close();
        saved++;
      } catch (err) {
        stderr.writeln('distill-hook: failed to save chunk: $err');
      }
    }

    stderr.writeln(
      'distill-hook: extracted ${chunks.length}, saved $saved knowledge chunks',
    );
  } catch (err) {
    stderr.writeln('distill-hook: extraction failed: $err');
    exit(1);
  }
}

Future<String> _readStdin() async {
  final buffer = StringBuffer();
  try {
    await for (final chunk in stdin.transform(utf8.decoder)) {
      buffer.write(chunk);
    }
  } catch (_) {
    // stdin closed or errored
  }
  return buffer.toString().trim();
}
